aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp

```diff
@@ -411,14 +454,10 @@ namespace Aws
                 m_encryptionMode = false;
                 m_decryptionMode = false;
 
-                if (m_ctx.cipher || m_ctx.cipher_data || m_ctx.engine)
-                {
-                    EVP_CIPHER_CTX_cleanup(&m_ctx);
-                }
-
-                m_ctx.cipher = nullptr;
-                m_ctx.cipher_data = nullptr;
-                m_ctx.engine = nullptr;
+#if OPENSSL_VERSION_NUMBER < 0x10100003L
+                EVP_CIPHER_CTX_cleanup(m_ctx);
+                m_ctx = nullptr;
```

Comment:
I think this is OK. The code paths are slightly different (I think it's a bit ugly and hard to follow, actually). The idea was to be able to reuse the same EVP_CIPHER_CTX throughout the lifetime of a class OpenSSLCipher rather then freeing and re-allocating to achieve a more static memory allocation modelling, comparable to the OpenSSL 1.0 implementation.

Cleanup() is called from the destructor and from ::Reset(). For OpenSSL 1.1:
- Cleanup() itself doesn't do anything to the EVP_CIPHER_CTX
- the destructor does Cleanup(), followed by EVP_CIPHER_CTX_free()
- ::Reset() does Cleanup(), followed by Init(), which itself does EVP_CIPHER_CTX_reset()
hence we achieve re-use of the handle, rather than free() + malloc()

For OpenSSL 1.0:
- Cleanup() itself invalidates the static handle as per your comment
- the destructor does Cleanup()
- ::Reset() does Cleanup(), followed by Init(), which itself reinitialises the static handle